### DISCUSSION TITLE
Best way to scale up number of PMTiles datasets served

### DISCUSSION THREAD

**Discussion by alexjball** [maplibre/martin#1125] [Q&A]
We would like to use Martin to serve datasets from PMTiles. We generate datasets based on user-initiated actions, and so would like to be able to dynamically update martin's source config to serve these datasets when available. We would have one martin deployment that serves from an arbitrary number of datasets. Another container would proxy requests to/from martin to handle auth from clients and to the source files. 

1. Is there a way to update the config without restarting Martin, or would we need to do a rolling update every time there is a new dataset available
2. Is there significant overhead to configuring 1 sources vs 10k sources?
3. Is there another approach you'd recommend for scaling up the number of datasets served? This could reduce the effectiveness of martin's tile caching, though a CDN could be used instead.

**Comment by nyurik** [comment]
Hi @alexjball, 

> 1. Is there a way to update the config without restarting Martin, or would we need to do a rolling update every time there is a new dataset available

This has been a long requested feature - see long thread in #288 -- would love to have it, but someone has to sit down and implement it. Not trivial, but doable.

> 2. Is there significant overhead to configuring 1 sources vs 10k sources?

PMTiles don't have much overhead - they all share the same cache system for their directory entries. If you have 10,000 pmtiles, they will all have to be open at the same time -- consuming OS resources (file handles), but shouldn't be too big of a problem... I could be wrong of course :)

> 3. Is there another approach you'd recommend for scaling up the number of datasets served? This could reduce the effectiveness of martin's tile caching, though a CDN could be used instead.

You can easily shard pmtiles serving - e.g. serve pmtiles that start from a-m from one server, and n-z from another.  But I wouldn't worry about these things until you actually see a problem - doubtful that having 10,000 files is that big of a deal tbh


P.S. when configuring pmtiles - you can simply point martin to a directory of pmtiles files - it will pick them all up

**Reply by alexjball** [reply]
Thank you for the guidance! We would like to host the tiles on S3. It looks like Martin fetches all the tilejson metadata at startup, so my biggest concern is startup latency building the catalog, and we won't need the catalog endpoint. But as you say, probably not an issue until it's an issue. 

S3 access is proxied so Martin sees it as a plain http endpoint. Does Martin support directory discovery for the http backend?

**Reply by nyurik** [reply]
Ah, thx, didn't realize you were using HTTP. At the moment, there is no directory support of any kind (kinda tricky to do, unless we want to do html scraping ... not pretty... but PRs are welcome :)   All HTTP connections are done via a single instance of a [reqwest](https://crates.io/crates/reqwest) crate, so if all of them are on the same server, it should be well optimized (i.e. connection sharing).

That said, we really need to add "proper" S3 support, so that Martin can:
* understand that its an S3 bucket
* use S3 authentication
* be able to enumerate all available files via standard S3 API

**Reply by alexjball** [reply]
I would be interested in working on s3 support, priority will be set based on our experience deploying via config updates. I am new to Rust so that's the main hurdle...

In case I get to this, I would think to create a new [impl_pmtiles_source](https://github.com/maplibre/martin/blob/main/martin/src/pmtiles/mod.rs#L287) using [rust-s3](https://github.com/durch/rust-s3), which would support s3-compatible api's. Is it acceptable to add that dependency?

**Reply by nyurik** [reply]
@alexjball thanks!  I suspect you may need to modify the pmtiles crate first (i'll be happy to review/merge it there) - https://github.com/stadiamaps/pmtiles-rs/blob/main/src/async_reader.rs 

Currently in pmtiles, reader has two backend implementations: `HttpBackend` and `MmapBackend`.  The `HttpBackend` is very similar to what we need for S3 access. Inside, it uses a popular [reqwest](https://crates.io/crates/reqwest) crate (similar to python's i believe) to establish and manage connections, and to get the needed data. Underneath, `reqwest` uses [hyper](https://crates.io/crates/hyper) - a ubiquitous low-level HTTP crate. Note that `hyper` actually recommends users to use `reqwest` for higher-level API.

The `rust-s3` crate also uses `hyper`, but it uses it directly (they used to depended on `reqwest`, but moved to hyper a few months ago in https://github.com/durch/rust-s3/pull/299) - I don't know if that was a good or a bad decision tbh -- I have never used hyper directly.

So the solution could be one of these:
* Add a new `S3Backend` that calls `Bucket::get_object_range` - almost identical to `HttpBackend`.  cc: @lseelenbinder 
* Same as first, but implement `S3Backend` in Martin directly. This could be a faster path, and might be good to get started, but eventually I would prefer to move this code to upstream lib - will benefit all other pmtiles users, and consolidates testing.

Once pmtiles backend is in place, you can create a new `PmtS3Source` similar to `PmtHttpSource`.

**Reply by lseelenbinder** [reply]
@nyurik I think we'll be looking into S3 support directly anyways, so this may happen sooner rather than later. 😄
